jobs_worker.go

```diff
@@ -0,0 +1,133 @@
+package foundation
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+
+	"github.com/gocraft/work"
+	"github.com/gomodule/redigo/redis"
+)
+
+const (
+	defaultConcurrency = 5
+	defaultNamespace   = "foundation_jobs_worker"
+)
+
+// JobsWorkerContext base context for workers to use
+type JobsWorkerContext struct {
+}
+
+type JobsWorker struct {
+	*Service
+
+	Options *JobsWorkerOptions
+}
+
+func InitJobsWorker(name string) *JobsWorker {
+	return &JobsWorker{
+		Service: Init(name),
+	}
+}
+
+type JobOptions struct {
+	handler  func(job *work.Job) error
+	schedule string
+	options  *work.JobOptions
+}
+
+// JobsWorkerOptions represents the options for starting a jobs worker
+type JobsWorkerOptions struct {
+	// JobHandlers are the handlers to use for the jobs
+	Jobs map[string]JobOptions
+	// JobMiddlewares are the middlewares to use for all jobs
+	Middlewares []func(job *work.Job, next work.NextMiddlewareFunc) error
+	// Namespace is the redis namespace to use for the jobs
+	Namespace string
+	// Concurrency is the number of concurrent jobs to run
+	Concurrency int
+	// StartComponentsOptions are the options to start the components.
+	StartComponentsOptions []StartComponentsOption
+}
+
+func NewJobsWorkerOptions() *JobsWorkerOptions {
+	return &JobsWorkerOptions{
+		Namespace:   defaultNamespace,
+		Concurrency: defaultConcurrency,
+	}
+}
+
+// Start runs the worker that handles jobs
+func (w *JobsWorker) Start(opts *JobsWorkerOptions) {
+	w.Options = opts
+
+	w.Service.Start(&StartOptions{
+		ModeName:               "jobs_worker",
+		StartComponentsOptions: w.Options.StartComponentsOptions,
+		ServiceFunc:            w.ServiceFunc,
+	})
+}
+
+func (w *JobsWorker) ServiceFunc(ctx context.Context) error {
+	redisUrl := GetEnvOrString("JOBS_REDIS_URL", "")
+	if redisUrl == "" {
+		return fmt.Errorf("JOBS_REDIS_URL is required")
+	}
+
+	var redisPool = &redis.Pool{
+		MaxActive: w.Options.Concurrency,
+		MaxIdle:   w.Options.Concurrency,
+		Wait:      true,
+		Dial: func() (redis.Conn, error) {
+			return redis.Dial("tcp", redisUrl)
+		},
+	}
+
+	workerPool := work.NewWorkerPool(JobsWorkerContext{}, uint(w.Options.Concurrency), w.Options.Namespace, redisPool)
+
+	workerPool.Middleware(w.Logger)
+
+	if w.Options.Middlewares != nil {
+		for _, middleware := range w.Options.Middlewares {
+			workerPool.Middleware(middleware)
+		}
+	}
+
+	for jobName, jobOptions := range w.Options.Jobs {
+		if jobOptions.handler == nil {
+			return fmt.Errorf("job %s has no handler", jobName)
+		}
+
+		if jobOptions.options != nil {
+			workerPool.JobWithOptions(jobName, *jobOptions.options, jobOptions.handler)
+		}
+
+		if jobOptions.schedule != "" {
+			workerPool.PeriodicallyEnqueue(jobOptions.schedule, jobName)
+		}
+	}
+
+	workerPool.Start()
+
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt)
+	<-signalChan
```

Comment:
You can just
```suggestion
	<-ctx.Done()
```

It will be the same

Reply:
Thanks, fixed